### PR TITLE
#9264 - EnergyPlus missing vcomp140.dll on Windows

### LINKED ISSUE #9264: EnergyPlus v9.6.0 on Windows 11 missing dlls
Issue overview
--------------
Installing EnergyPlus v9.6 on a new Windows 11 machine results in errors from missing DLLs, e.g. vcomp140.dll

Workaround:
Installing the [Visual C++ Redistributable library](https://docs.microsoft.com/en-US/cpp/windows/latest-supported-vc-redist?view=msvc-170) should fix things.
Specifically this version: https://aka.ms/vs/17/release/vc_redist.x64.exe

### Details
Some additional details for this issue (if relevant):
 - Platform Windows 11
 - Version of EnergyPlus 9.6.0
 - [Unmethours link](https://unmethours.com/question/68651/dll-errors-installing-os-and-e-on-windows-11/)

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here) n/a
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------

- Fix #9264 
- EnergyPlus missing vcomp140.dll on Windows. This happens when something links to OpenMP.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: cmake/Install.cmake [hunk 1]
```diff
@@ -664,6 +664,17 @@ if(WIN32 AND NOT UNIX)
...
+
+  # ConvertInputFormat is added via add_subdirectory, and in there we set CMAKE_INSTALL_OPENMP_LIBRARIES at parent scope already
+  # Otherwise, if either cpgfunctionEP (yes by default) or kiva (no by default) **actually** linked to OpenMP, we need to ship the libs
+  if (NOT ${CMAKE_INSTALL_OPENMP_LIBRARIES} AND (${USE_OpenMP} OR ${ENABLE_OPENMP}))
+    # Need to install vcomp140.dll or similar
+    find_package(OpenMP COMPONENTS CXX)
+    if(OpenMP_CXX_FOUND)
+      set(CMAKE_INSTALL_OPENMP_LIBRARIES TRUE)
+    endif()
+  endif()
```
Comment (jmarrec) [inline cmake/Install.cmake:676]: this is the actual fix, trying to cover all cases (but really ConvertInputFormat should have set `CMAKE_INSTALL_OPENMP_LIBRARIES` if openMP was found)

File: src/ConvertInputFormat/CMakeLists.txt [hunk 2]
```diff
@@ -9,8 +9,9 @@ endif()
 
 # Detect OpenMP support in a compiler. If the compiler supports OpenMP, the
 # flags to compile with OpenMP are returned and added.
-find_package(OpenMP)
-if(OPENMP_FOUND)
+find_package(OpenMP COMPONENTS CXX)
+if(OpenMP_CXX_FOUND)
+  set(CMAKE_INSTALL_OPENMP_LIBRARIES TRUE PARENT_SCOPE)
```
Comment (jmarrec) [inline src/ConvertInputFormat/CMakeLists.txt:14]: In combination with this.

File: cmake/qtifw/install_win_copydll.qs [hunk 3]
```diff
@@ -50,9 +51,9 @@ function Component()
 
       // Store ocx to be registered
       var dllsToReg = [];
-      for (var i = 0; i < systemArray.length; i++) {
-        var sourceFile = `${tempDir}\\${systemArray[i]}`;
-        var targetFile = systemTargetDir + systemArray[i];
+      systemArray.forEach(dll => {
+        var sourceFile = `${tempDir}\\${dll}`;
+        var targetFile = `${systemTargetDir}\\${dll}`;
```
Comment (jmarrec) [inline cmake/qtifw/install_win_copydll.qs:56]: The rest is just clean up in the scripts to make @Myoldmopar happy with the use of forEach loops and co

**Review by jmarrec — COMMENTED** [review]
I tested with https://github.com/jmarrec/EnergyPlus/releases/tag/9264_Test2 that it will correctly install and can run a simulation in a clean windows 11 VM (after confirming the initial bug in a clean windows 11: vcomp140.dll was missing)

The installed folder does have vcomp140.dll next to the energyplus.exe now

**Comment by mjwitte** [comment]
@JasonGlazer (or anyone else) can you test this on a clean Windows *10* machine to make sure these changes don't break anything?

**Comment by JasonGlazer** [comment]
@mjwitte I should be able to create a clean Windows 10 VM and test it.

**Comment by JasonGlazer** [comment]
@mjwitte I was able to run energyplus.exe --help and convertinputformat.exe --help on a clean Windows 10 machine without a problem. I also ran EP-Launch and run a few different example files and they all worked. No complaints about missing dll's.

**Comment by mjwitte** [comment]
@JasonGlazer Thanks. And IDF Editor and VersionUpdater?

**Comment by JasonGlazer** [comment]
@mjwitte yes, they work fine.

**Review by mjwitte — APPROVED** [review]
Based on @jmarrec and @JasonGlazer 's testing, this is going in.

**Comment by mjwitte** [comment]
@LipingWang Windows 11 installer is fixed.

**Comment by mjwitte** [comment]
Arrgh. This is failing on [alternate_build_configurations (ubuntu-20.04)](https://github.com/NREL/EnergyPlus/runs/5560274282?check_suite_focus=true#logs)
```
Reading package lists...
W: Conflicting distribution: https://packages.microsoft.com/ubuntu/20.04/prod focal InRelease (expected focal but got testing)
E: Repository 'https://packages.microsoft.com/ubuntu/20.04/prod focal InRelease' changed its 'Origin' value from 'microsoft-ubuntu-focal-prod focal' to 'microsoft-ubuntu-focal-prod testing'
E: Repository 'https://packages.microsoft.com/ubuntu/20.04/prod focal InRelease' changed its 'Label' value from 'microsoft-ubuntu-focal-prod focal' to 'microsoft-ubuntu-focal-prod testing'
E: Repository 'https://packages.microsoft.com/ubuntu/20.04/prod focal InRelease' changed its 'Codename' value from 'focal' to 'testing'
Error: Process completed with exit code 100.
```
Any suggestions?

**Comment by mitchute** [comment]
> Arrgh. This is failing on [alternate_build_configurations (ubuntu-20.04)](https://github.com/NREL/EnergyPlus/runs/5560274282?check_suite_focus=true#logs)

I restarted the build. It was just a CI fluke. Looks like it's proceeding.

**Comment by JasonGlazer** [comment]
BTW, I just tested the installer on a Windows 11 VM and it worked fine.

**Pull request merged**